server/build_event_protocol/build_event_handler/build_event_handler.go

```diff
@@ -212,6 +212,27 @@ func (e *EventChannel) HandleEvent(ctx context.Context, event *pepb.PublishBuild
 		}
 	}
 
+	// When we get the workspace status event, update the invocation in the DB
+	// so that it can be searched by its commit SHA, user name, etc. even
+	// while the invocation is still in progress.
+	if isWorkspaceStatusEvent(bazelBuildEvent) {
+		db := e.env.GetInvocationDB()
+		// TODO: Make sure the lookup/insert happens in a single transaction.
+		// This isn't an issue currently since all invocation-modifying events
+		// are coming through the PublishBuildEvent stream, but using a transaction
+		// here would make this more future-proof.
+		ti, err := db.LookupInvocation(ctx, iid)
+		if err != nil {
+			return err
+		}
+		invocationProto := TableInvocationToProto(ti)
+		event_parser.FillInvocationFromWorkspaceStatus(bazelBuildEvent.GetWorkspaceStatus(), invocationProto)
+		ti = tableInvocationFromProto(invocationProto, ti.BlobID)
+		if err := db.InsertOrUpdateInvocation(ctx, ti); err != nil {
```

Comment:
What happens if all 3 events that trigger a write come in at the same time (the initial event, the workspace status event, and the final event). Is there room for race conditions in these 3 mysql writes? Seems like it'd be pretty common for cached builds.

Reply:
AFAIK all events for an invocation come across the same stream -- if they come in three separate build events on that stream, then I think this is OK?

(Brandon -- can you double check and convince yourself that this is the case before we go ahead with this?)

Reply:
Yeah, the loop looks like

```
for {
  in, ... := stream.Recv()
  ...  
  channel.HandleEvent(ctx, in)
}
```

And `Recv` / `HandleEvent` are blocking (IIUC). As long as those are the only things that come in and modify the invocation while it's in progress, I think we should be OK.

One corner case I can think of though: maybe someone could try to change the sharing settings on an invocation while it's in progress? In that case we might want to prevent mutating an invocation while the status is in progress (except when finalizing it)